src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessFetch.cs

```diff
@@ -54,13 +55,18 @@ public void Process(FetchRequestBase resultOperator, QueryModelVisitor queryMode
 													.GetClassMetadata(resultOperator.RelationMember.ReflectedType);
 					if (metadata == null)
 					{
-						var entityName = queryModelVisitor.VisitorParameters.SessionFactory.GetImplementors(
-							resultOperator.RelationMember.ReflectedType.FullName).FirstOrDefault();
-						if (!string.IsNullOrEmpty(entityName))
+						foreach (var entityName in queryModelVisitor.VisitorParameters.SessionFactory
+							         .GetImplementors(resultOperator.RelationMember.ReflectedType.FullName))
```

Comment:
The main problem here that we don't know the actual entity type. Only declaring type of fetched property (which is not equal to entity type if this property is from some base class). That's why we need to check all implementors.

Reply:
https://github.com/nhibernate/nhibernate-core/pull/3320#issuecomment-1595644278

Reply:
> The main problem here that we don't know the actual entity type. 

Actually we do. It is in query source as you've said.